module/helper/renderer/src/webgl/program.rs

```diff
@@ -41,64 +41,118 @@
 mod private
 {
   use minwebgl as gl;
-  use std::collections::HashMap;
+  use gl::GL;
+  use web_sys::WebGlProgram;
+  use rustc_hash::FxHashMap;
 
+  /// Shader program generalization for getting access to used shader locations
+  pub trait ShaderProgram : clone_dyn_types::CloneDyn
+  {
+    /// Returns a reference to the hash map containing uniform locations
+    fn get_locations( &self, gl : &GL, program : &gl::WebGlProgram ) -> FxHashMap< String, Option< gl::WebGlUniformLocation > >;
+
+    /// Returns a reference to the hash map containing UBO indices
+    fn get_ubo_indices( &self, gl : &GL, program : &gl::WebGlProgram ) -> FxHashMap< String, u32 >;
+
+    /// Get [`ShaderProgram`] type name
+    fn get_type_name( &self ) -> &'static str;
+
+    /// Create new boxed clone of self
+    fn dyn_clone( &self ) -> Box< dyn ShaderProgram >;
+  }
```

Comment:
Use a newtype instead
Each new Shader is going to be a wrapper aorung ProgramInfo. ShaderProgram trait is going to be implemented for this newtype. You can add a branch to impl_locations can create this new type by itself, so the user does not have to